File: app/lib/shared/models/order_state_machine.dart
import 'package:foodhome_app/shared/models/order_status.dart';

bool canTransitionOrder({
  required OrderStatus from,
  required OrderStatus to,
}) {
  return nextOrderStatuses(from).contains(to);
}

List<OrderStatus> nextOrderStatuses(OrderStatus status) {
  return switch (status) {
    OrderStatus.requested => [
      OrderStatus.accepted,
      OrderStatus.cancelled,
    ],
    OrderStatus.accepted => [
      OrderStatus.cooking,
      OrderStatus.blocked,
      OrderStatus.cancelled,
    ],
    OrderStatus.cooking => [
      OrderStatus.blocked,
      OrderStatus.served,
    ],
    OrderStatus.blocked => [
      OrderStatus.accepted,
      OrderStatus.cooking,
      OrderStatus.cancelled,
    ],
    OrderStatus.served || OrderStatus.cancelled => const [],
  };
}

File: functions/src/domain/order_state_machine.ts
export const orderStatuses = [
  "requested",
  "accepted",
  "cooking",
  "blocked",
  "served",
  "cancelled",
] as const;

export type OrderStatus = (typeof orderStatuses)[number];

const nextStatusMap = {
  requested: ["accepted", "cancelled"],
  accepted: ["cooking", "blocked", "cancelled"],
  cooking: ["blocked", "served"],
  blocked: ["accepted", "cooking", "cancelled"],
  served: [],
  cancelled: [],
} satisfies Record<OrderStatus, readonly OrderStatus[]>;

export function nextOrderStatuses(status: OrderStatus): readonly OrderStatus[] {
  return nextStatusMap[status];
}

export function canTransitionOrder(
  from: OrderStatus,
  to: OrderStatus,
): boolean {
  const nextStatuses = nextStatusMap[from] as readonly OrderStatus[];
  return nextStatuses.includes(to);
}
